### PR TITLE
Feat/selection methods

### PR DESCRIPTION
**Pull request by rosecers**
Subject to #3, this is a PR to merge the selection methods into sklearn-cosmo. 

As with PCovR, `alpha` has been renamed `mixing` to avoid confusion with sklearn nomenclature.

This PR will contain three things:
 - [X] code for selection methods 
     - [X] `CUR`, PCov-CUR which defaults to CUR when mixing not specified or set to 1
     - [X] `FPS`, PCov-CUR which defaults to FPS when mixing not specified or set to 1
 - [X] docs for selection methods 
     - [X] `CUR`, PCov-CUR which defaults to CUR when mixing not specified or set to 1
     - [X] `FPS`, PCov-CUR which defaults to FPS when mixing not specified or set to 1
 - [x] examples
     - [X] feature selection example notebooks
     - [ ] ~~sample selection example notebooks~~ (this cannot be added until sparse methods are pulled into skcosmo)
 - [ ] tests
     - [ ] catching convergence errors
     - [ ] centering?

Can we review this in multiple stages, so that by the time the examples/tests are written, we've converged on the code? I'd like to add them all to main together.

### CODE REVIEW

**Comment by Luthaf** [comment]
> feature selection example notebooks

If we want to use notebooks for examples (which I agree is nice), how can we make sure we don't have the usual issues of having random changes in them to review in PR?

We have two options: either we try to maintain the notebook in a "clean" state all the time, for which nbstripout is part of the answer, but not everything (for example it does not remove the "last python version used" field); or we don't care, and ignore these changes when reviewing.

If we are using notebooks for examples, I would actually tend to prefer leaving the output so that just viewing the notebook in github web interface or with binder already show all plots. We would still have to be a bit careful with the notebook size. 

A third alternative is to put the example notebooks in kernel-tutorials instead of this repo. We can still have plain python examples here.

File: skcosmo/selection/CUR.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
+from abc import abstractmethod
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:1]: This module needs a module docstring

File: skcosmo/selection/CUR.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+import numpy as np
+import scipy
+from scipy.sparse.linalg import eigs as speig
+
+from ._base import _BaseSelection
+
+class _CUR(_BaseSelection):
+    """
+    Super-class defined for CUR selection methods
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:10]:
```suggestion
    Base class for CUR selection methods
```

Why do we need a base class here?

File: skcosmo/selection/CUR.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+    Parameters
+    ----------
+    mixing : float
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:14]: How is that related to CUR?

File: skcosmo/selection/CUR.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        whether to orthogonalize the matrices after each selection
+    k : int
+        number of eigenvectors to compute the importance score with
+    matrix : ndarray of shape (n x m)
+        Data to select from -
+        Feature selection will choose a subset of the `m` columns
+        Samples selection will choose a subset of the `n` rows
+        s   tored in `self.A`
+    precompute : int
+        Number of selections to precompute
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:27]: that is not very clear. The docstring needs a high level overview of what is going on for users to be able to understand these parameters

File: skcosmo/selection/CUR.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    progress_bar : bool, callable
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:28]: This feels like scope creep: why should this be part of CUR? Is this or something similar used in sklearn? 

If anything, I would move progress report to a separate class used as a mixin.

File: skcosmo/selection/CUR.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        Option to include tqdm progress bar or a callable progress bar
+        implemented in `self.progress(iterable)`
+    tolerance : float
+        Threshold below which values will be considered 0
+        stored in `self.tol`
+    Y (optional) : ndarray of shape (n x p)
+        Array to include in biased selection when mixing < 1
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:35]: This should not be part of the base class, as for mixin above, but defined in the PCovCUR derived class

File: skcosmo/selection/CUR.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
+from abc import abstractmethod
+import numpy as np
+import scipy
+from scipy.sparse.linalg import eigs as speig
+
+from ._base import _BaseSelection
+
+class _CUR(_BaseSelection):
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:8]:
```suggestion
class CURBase(_BaseSelection):
```

File: skcosmo/selection/_base.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
+from abc import abstractmethod
+
+
+class _BaseSelection:
```
Comment (Luthaf) [inline skcosmo/selection/_base.py:4]: I'm not completely sold on using deep class hierarchy here. Does sklearn do this or something similar? 

In Python, class inheritance is mostly useful to provide implementation of common methods, which is not the case here. The only method implemented is `progress`, which is not related to the core of what a "Selection method" is.

File: skcosmo/selection/FPS.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+            where ~Y is the properties obtained by linear regression.
+        """
+
+        C = np.zeros((self.A.shape[1], self.A.shape[1]), dtype=np.float64)
+
+        cov = self.A.T @ self.A
+
+        if self.mixing < 1:
+            # changing these next two lines can cause a LARGE error
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:154]: Could you explain why as part of the comment as well?

File: skcosmo/selection/FPS.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        K = np.zeros((self.A.shape[0], self.A.shape[0]))
+        if self.mixing < 1:
+            K += (1 - self.mixing) * self.Y @ self.Y.T
+        if self.mixing > 0:
+            K += (self.mixing) * self.A @ self.A.T
+
+        return K
+
+
+class featureFPS(_FPS):
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:135]: This needs a docstring

File: skcosmo/selection/FPS.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        return self.idx
+
+    @abstractmethod
+    def get_product(self):
+        """Abstract method for computing the product (C or K) of the input matrices
+        """
+        return
+
+
+class sampleFPS(_FPS):
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:114]: this needs a docstring

File: skcosmo/selection/FPS.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
+from abc import abstractmethod
+import numpy as np
+import scipy
+
+from ._base import _BaseSelection
+
+def calculate_pcov_distances(points, ref_idx, idxs=None):
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:7]: this needs a docstring, and looks like a private function, so it should be named `_calculate_pcov_distances`

**Review by Luthaf — COMMENTED** [review]
I had a quick look, and I would like to discuss a bit the use of class hierarchy. I feel like we don't get much by having it as deep as it currently is. I think a `BaseFPS` and `BaseCUR` should be more than enough here.

**Comment by Luthaf** [comment]
Can you rebase this on top of the main branch? This way the tests will run in github actions!

File: skcosmo/selection/CUR.py (hunk 3 above)
Comment (rosecers) [reply skcosmo/selection/CUR.py:14]: This is for PCov-CUR, which defaults to CUR when mixing = 1

File: skcosmo/selection/FPS.py (hunk 9 above)
Comment (rosecers) [reply skcosmo/selection/FPS.py:154]: If you try to make it more efficient by computing few eigenvalues of `C`, the resulting selections will not be correct

**Comment by rosecers** [comment]
@Luthaf can we merge as-is and I'll create a new one for further tests/examples?

File: examples/FeatureSelection.ipynb [hunk 13]
```diff
@@ -0,0 +1,172 @@
+{
```
Comment (Luthaf) [inline examples/FeatureSelection.ipynb:1]: It would be nice to render this in the documentation directly. Do you know if it is possible?

File: skcosmo/pcovr/pcovr_distances.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
+import numpy as np
+import scipy
+
+
+def get_Ct(mixing, X_proxy, Y_proxy, rcond):
```
Comment (Luthaf) [inline skcosmo/pcovr/pcovr_distances.py:5]:
```suggestion
def pcovr_covariance(mixing, X_proxy, Y_proxy, rcond):
```

File: skcosmo/pcovr/pcovr_distances.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    Creates the PCovR modified covariance
```
Comment (Luthaf) [inline skcosmo/pcovr/pcovr_distances.py:7]: This is missing parameters documentation

File: skcosmo/pcovr/pcovr_distances.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+        C += (1 - mixing) * Y_hat @ Y_hat.T
+
+    if mixing > 0:
+        C += (mixing) * cov
+
+    return C
+
+
+def get_Kt(mixing, X_proxy, Y_proxy, rcond):
```
Comment (Luthaf) [inline skcosmo/pcovr/pcovr_distances.py:36]:
```suggestion
def pcovr_kernel_distance(mixing, X_proxy, Y_proxy, rcond):
```

File: skcosmo/pcovr/pcovr_distances.py [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    Creates the PCovR modified kernel distances
```
Comment (Luthaf) [inline skcosmo/pcovr/pcovr_distances.py:38]: This is missing parameters documentation

File: skcosmo/selection/CUR.py (hunk 1 above)
Comment (Luthaf) [reply skcosmo/selection/CUR.py:1]: Still true!

File: skcosmo/selection/CUR.py [hunk 18]
```diff
@@ -0,0 +1,297 @@
+from abc import abstractmethod
+import numpy as np
+from scipy.sparse.linalg import eigs as speig
+
+from skcosmo.pcovr.pcovr_distances import get_Ct, get_Kt
+from .orthogonalizers import feature_orthogonalizer, sample_orthogonalizer
+
+
+class _CUR:
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:9]:
```suggestion
class BaseCUR:
```

or 

```suggestion
class CURBase:
```

File: skcosmo/selection/CUR.py [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+    """
+    Base class for CUR selection methods
+    Requires a product, typically the gram or covariance matrix, \
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:12]:
```suggestion
    Requires a product, typically the gram or covariance matrix,
```

File: skcosmo/selection/CUR.py [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+                      defaults to `true`
+    :type iterative: boolean
+
+    :param k: number of eigenvectors to compute the importance score with,
+              defaults to 1
+    :type k: int
+
+    :param tolerance: threshold below which values will be considered 0,
+                      defaults to 1E-12
+    :type tolerance: float0
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:29]: Should this be

```suggestion
    :type tolerance: float
```

File: skcosmo/selection/CUR.py [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+    """
+
+    def __init__(self, iterative=True, tolerance=1e-12, k=1, **kwargs):
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:32]: There is no need to accept kwargs if we are doing nothing with them

```suggestion
    def __init__(self, iterative=True, tolerance=1e-12, k=1):
```

File: skcosmo/selection/CUR.py [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+
+        self.k = k
+        self.iter = iterative
+        self.tol = tolerance
+        self.idx = []
+        self.pi = []
+
+    def select(self, n):
+        """
+        Method for CUR select based upon a product of the input matrices
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:42]: That does not help me understand the code. I find such API documentation much easier to understand if the docstring tells me **what** the function does. Something like `Select n entries from the product of input matrix ...`

File: skcosmo/selection/CUR.py [hunk 23]
```diff
@@ -0,0 +1,297 @@
...
+
+        :param n: number of selections to make
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:44]:
```suggestion
        :param n: requested number of selections. This function might return less selections if ???
```

File: skcosmo/selection/CUR.py [hunk 24]
```diff
@@ -0,0 +1,297 @@
...
+        :returns: list of selected indices
+        :rtype: list of int
+
+        """
+
+        if len(self.idx) > n:
+            return self.idx[:n]
+
+        for i in range(len(self.idx), n):
+            # try:
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:56]: leftover

File: skcosmo/selection/CUR.py [hunk 25]
```diff
@@ -0,0 +1,297 @@
...
+        """
+        return
+
+    @abstractmethod
+    def orthogonalize(self):
+        """Method to orthogonalize matrices"""
+        return
+
+
+class sampleCUR(_CUR):
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:91]: Python uses CamelCase for class name

```suggestion
class SampleCUR(_CUR):
```

File: skcosmo/selection/CUR.py [hunk 26]
```diff
@@ -0,0 +1,297 @@
...
+    :type tolerance: float
+
+    :param Y: array to include in biased selection when mixing < 1; required
+              when mixing < 1, throws AssertionError otherwise
+    :type Y: array of shape (n x p), optional
+
+
+    """
+
+    def __init__(self, matrix, mixing=1.0, iterative=True, tolerance=1e-12, **kwargs):
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:134]: Same as above re kwargs.

File: skcosmo/selection/CUR.py [hunk 27]
```diff
@@ -0,0 +1,297 @@
...
+            \\mathbf{X}_\\mathbf{r}^T \\mathbf{Y}_\\mathbf{r}}`.
+
+        """
+        if self.iter:
+            self.A_current, self.Y_current = sample_orthogonalizer(
+                self.idx, self.A_current, self.Y_current, self.tol
+            )
+
+
+class featureCUR(_CUR):
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:197]:
```suggestion
class FeatureCUR(_CUR):
```

File: skcosmo/selection/CUR.py [hunk 28]
```diff
@@ -0,0 +1,297 @@
...
+                      defaults to 1E-12
+    :type tolerance: float
+
+    :param Y: array to include in biased selection when mixing < 1;
+              required when mixing < 1, throws AssertionError otherwise
+    :type Y: array of shape (n x p), optional
+    """
+
+    def __init__(
+        self, matrix, mixing=1.0, iterative=True, tolerance=1e-12, k=1, **kwargs
```
Comment (Luthaf) [inline skcosmo/selection/CUR.py:240]: kwargs

File: skcosmo/selection/FPS.py [hunk 29]
```diff
@@ -0,0 +1,171 @@
+import numpy as np
+from skcosmo.pcovr.pcovr_distances import get_Ct, get_Kt
+
+
+def _calculate_distances_(points, ref_idx, idxs=None):
+    if idxs is None:
+        idxs = range(points.shape[0])
+    return np.array(
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:8]: I'm not sure creating a list using list comprehension and then converting it to a numpy array is the most efficient thing to do. It should be fine for now, but if there is a way to transform this to using only numpy operations it would be nice

File: skcosmo/selection/FPS.py [hunk 30]
```diff
@@ -0,0 +1,171 @@
...
+        [
+            np.real(points[j][j] - 2 * points[j][ref_idx] + points[ref_idx][ref_idx])
+            for j in idxs
+        ]
+    )
+
+
+class _FPS:
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:16]:
```suggestion
class BaseFPS:
```

```suggestion
class FPSBase:
```

File: skcosmo/selection/FPS.py [hunk 31]
```diff
@@ -0,0 +1,171 @@
...
+    """
+    Base Class defined for FPS selection methods
+
+    :param idxs: predetermined indices; if None provided, first index selected
+                 is random
+    :type idxs: list of int, None
+
+    """
+
+    def __init__(self, idxs=None, **kwargs):
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:26]: kwargs

File: skcosmo/selection/FPS.py [hunk 32]
```diff
@@ -0,0 +1,171 @@
...
+        else:
+            self.idx = [np.random.randint(self.product.shape[0])]
+        self.distances = np.min(
+            [_calculate_distances_(self.product, idx) for idx in self.idx], axis=0
+        )
+        self.distance_selected = np.nan * np.zeros(self.product.shape[0])
+
+    def select(self, n):
+        """
+        Method for FPS select based upon a product of the input matrices
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:38]: same as BaseCUR, this is not very clear to me.

File: skcosmo/selection/FPS.py [hunk 33]
```diff
@@ -0,0 +1,171 @@
...
+
+    def __init__(self, idxs=None, **kwargs):
+        if idxs is not None:
+            self.idx = idxs
+        else:
+            self.idx = [np.random.randint(self.product.shape[0])]
+        self.distances = np.min(
+            [_calculate_distances_(self.product, idx) for idx in self.idx], axis=0
+        )
+        self.distance_selected = np.nan * np.zeros(self.product.shape[0])
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:34]:
```suggestion
        self.distance_selected =  np.full(shape=self.product.shape[0], fill_value=np.nan)
```

File: skcosmo/selection/FPS.py [hunk 34]
```diff
@@ -0,0 +1,171 @@
...
+                )
+            self.idx.append(np.argmax(self.distances))
+            self.distance_selected[self.idx[-1]] = self.distances[self.idx[-1]]
+
+            if np.abs(self.distances).max() < self.tol:
+                return self.idx
+        return self.idx
+
+
+class sampleFPS(_FPS):
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:64]:
```suggestion
class SampleFPS(_FPS):
```

File: skcosmo/selection/FPS.py [hunk 35]
```diff
@@ -0,0 +1,171 @@
...
+    def __init__(self, matrix, mixing=1.0, tolerance=1e-12, **kwargs):
+
+        self.mixing = mixing
+        self.tol = tolerance
+
+        self.A = matrix.copy()
+
+        if mixing < 1:
+            try:
+                assert "Y" in kwargs
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:112]: I would rather use a `Y=None` parameter, it will show up both in documentation and tooltips in code editors.

File: skcosmo/selection/FPS.py [hunk 36]
```diff
@@ -0,0 +1,171 @@
...
+                self.Y = kwargs.get("Y")
+            except AssertionError:
+                print(r"For $\mixing < 1$, $Y$ must be in" "the constructor parameters")
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:115]: This should be an error, right? There is nothing we can do to save it here

```suggestion
                raise Exception(r"For $\mixing < 1$, $Y$ must be in" "the constructor parameters")
```

File: skcosmo/selection/FPS.py [hunk 37]
```diff
@@ -0,0 +1,171 @@
...
+        else:
+            self.Y = None
+
+        self.product = get_Kt(self.mixing, self.A, self.Y, rcond=self.tol)
+        super().__init__(tolerance=tolerance, **kwargs)
+
+
+class featureFPS(_FPS):
```
Comment (Luthaf) [inline skcosmo/selection/FPS.py:123]:
```suggestion
class FeatureFPS(_FPS):
```

File: skcosmo/selection/orthogonalizers.py [hunk 38]
```diff
@@ -0,0 +1,41 @@
+import numpy as np
+
+
+def feature_orthogonalizer(idx, A_proxy, Y_proxy, tol=1e-12):
```
Comment (Luthaf) [inline skcosmo/selection/orthogonalizers.py:4]: needs a docstring

File: skcosmo/selection/orthogonalizers.py [hunk 39]
```diff
@@ -0,0 +1,41 @@
...
+        v = A_proxy[:, idx[-1]] / np.sqrt(
+            np.matmul(A_proxy[:, idx[-1]], A_proxy[:, idx[-1]])
+        )
+
+        for i in range(A_proxy.shape[1]):
+            A_proxy[:, i] -= v * np.dot(v, A_proxy[:, i])
+    return A_proxy, Y_proxy
+
+
+def sample_orthogonalizer(idx, A_proxy, Y_proxy, tol=1e-12):
```
Comment (Luthaf) [inline skcosmo/selection/orthogonalizers.py:24]: needs a docstring

**Review by Luthaf — COMMENTED** [review]
The code looks good overall, I have a few suggestions to improve naming and default parameters. 

Regarding code organisation, this is missing a `skcosmo/selection/__init__.py` and `skcosmo/pcovr/__init__.py`. I think I would like to directly do `from skcosmo.selection import SampleFPS, FeatureCUR` instead of going all the way to `skcosmo.selection.CUR` and `skcosmo.selection.FPS`

This is still missing tests, if not very elaborate ones at least simple tests on sample sklearn datasets to prevent regression.

File: docs/source/index.rst [hunk 40]
```diff
@@ -1,9 +1,21 @@
...
+   :caption: Contents:
+
+   selection.rst
 
 ``sklearn-cosmo`` is a collection of `scikit-learn <https://sklearn.org/>`_
 compatible utilities that implement methods developed in the `COSMO laboratory
 <https://cosmo.epfl.ch>`_.
+A collection of scikit-learn compatible utilities that implement methods
+developed in the COSMO laboratory.
+
```
[the remark this answers is not in the document]
Comment (rosecers) [reply docs/source/index.rst]: Done!

File: skcosmo/selection/FPS.py (hunk 35 above)
Comment (rosecers) [reply skcosmo/selection/FPS.py]: Agreed

File: skcosmo/selection/CUR.py (hunk 1 above)
Comment (rosecers) [reply skcosmo/selection/CUR.py:1]: Done!

File: examples/FeatureSelection.ipynb (hunk 13 above)
Comment (rosecers) [reply examples/FeatureSelection.ipynb:1]: Hm, I'd have to look into it

File: skcosmo/selection/orthogonalizers.py (hunk 38 above)
Comment (rosecers) [reply skcosmo/selection/orthogonalizers.py]: done!

File: skcosmo/selection/orthogonalizers.py (hunk 39 above)
Comment (rosecers) [reply skcosmo/selection/orthogonalizers.py]: done!

File: skcosmo/selection/CUR.py (hunk 1 above)
Comment (Luthaf) [reply skcosmo/selection/CUR.py:2]: I meant a module-level docstring for "CUR.py", not a comment for the import 😃

Comment (Luthaf) [reply skcosmo/selection/CUR.py:2]: Like a description of what is in this file, and why

**Comment by rosecers** [comment]
> The code looks good overall, I have a few suggestions to improve naming and default parameters.
> 
> Regarding code organisation, this is missing a `skcosmo/selection/__init__.py` and `skcosmo/pcovr/__init__.py`. I think I would like to directly do `from skcosmo.selection import SampleFPS, FeatureCUR` instead of going all the way to `skcosmo.selection.CUR` and `skcosmo.selection.FPS`
> 
> This is still missing tests, if not very elaborate ones at least simple tests on sample sklearn datasets to prevent regression.

Hey, I think I got everything except the example rendering (which can wait IMO). Thoughts?

File: skcosmo/selection/CUR.py (hunk 1 above)
Comment (rosecers) [reply skcosmo/selection/CUR.py:2]: ... that's very confusing. on it.

File: skcosmo/pcovr/__init__.py [hunk 41]
```diff
@@ -0,0 +1,9 @@
+"""
+The :mod:`sklearn.selection` module includes FPS and CUR selection, each
```
Comment (Luthaf) [inline skcosmo/pcovr/__init__.py:2]:
```suggestion
The :mod:`skcosmo.selection` module includes FPS and CUR selection, each
```

File: tests/selection.py [hunk 42]
```diff
@@ -0,0 +1,53 @@
+import unittest
+import skcosmo
+from sklearn.datasets import load_boston
+
+
+class PCovSelectionTest:
+    def test_no_Y(self):
+        X, Y = load_boston(return_X_y=True)
+
+        with self.assertRaises(Exception):
```
Comment (Luthaf) [inline tests/selection.py:10]: You can check the message (to ensure that the error the test see is the one you think the test see) while still using the context manager version of `assertRaises` with 
```suggestion
        with self.assertRaises(Exception) as cm:
             # XXX
         self.assertEqual(str(cm.exception), "the error message")
```

File: skcosmo/selection/orthogonalizers.py [hunk 43]
```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
+"""
+
+This module contains the necessary orthogonalizers for the CUR decomposition
+subselection method.
+
+"""
+
+# Authors: Rose K. Cersonsky
```
Comment (Luthaf) [inline skcosmo/selection/orthogonalizers.py:9]: I would put the authors in the module docstring instead, so that it is rendered in the documentation. I don't think there is a need to specify the license, since it is the same one as the full package.

**Comment by Luthaf** [comment]
> Hey, I think I got everything except the example rendering (which can wait IMO). Thoughts?

I agree for the example rendering 😄 

There are a lot of unaddressed comments from my previous round of review though (hidden by github in the discussion tab/thread). You can see the full list here: https://github.com/cosmo-epfl/sklearn-cosmo/pull/4/files or by clicking "hidden conversation". They are mostly related to naming improvements and the use of kwargs.

File: skcosmo/selection/CUR.py (hunk 18 above)
Comment (rosecers) [reply skcosmo/selection/CUR.py:24]: I had intended that _CUR and _FPS never be called directly, hence the `_`. Would you be good with `_BaseCUR`?

Comment (Luthaf) [reply skcosmo/selection/CUR.py:24]: Sure, that's better!

File: tests/selection.py (hunk 42 above)
Comment (Luthaf) [reply tests/selection.py]: If the error come from code you wrote, I would say yes!

File: skcosmo/pcovr/__init__.py [hunk 44]
```diff
@@ -0,0 +1,9 @@
+"""
+The :mod:`skcosmo.selection` module includes FPS and CUR selection, each
```
Comment (Luthaf) [inline skcosmo/pcovr/__init__.py:2]: this is `skcosmo.pcovr`, not `skcosmo.selection`

File: skcosmo/selection/orthogonalizers.py (hunk 43 above)
Comment (Luthaf) [reply skcosmo/selection/orthogonalizers.py:9]: This one is still a comment

**Review by Luthaf — COMMENTED** [review]
Two very minor comments, and then I think this will be good to go.

We will have to figure out what is going on with the git history, I can have a look at this myself.

File: docs/source/selection.rst [hunk 45]
```diff
@@ -0,0 +1,76 @@
+Data Sub-Selection
```
Comment (Luthaf) [inline docs/source/selection.rst:1]: This file is not included in the main doctree

Comment (rosecers) [reply docs/source/selection.rst:1]: Done

**Review by Luthaf — APPROVED** [review]
Looks good! There is one remaining conflict with main to be fixed
